logger/src/errors.rs

```diff
@@ -0,0 +1,54 @@
+//! This module contains error handling and logging primitives.
+
+use crate::{debug, error, warn};
+
+/// Trait used for monadic error logging.
+pub trait ErrorLogging {
+    /// Log a warning using `self`, return `self`.
+    /// Examples:
+    /// `some_function_that_returns_error().warn("The error will be logged as a warning")?;`
+    fn log_warn(self, message: &str) -> Self;
```

Comment:
Do we have many cases in our code base where for an error we need to log it and still return it?

As far as I remember we usually either log it or return it - rarely both together.
But maybe there are use cases here, that need this way of handling. Generally I think it would be more useful to have a trait that just logs when there is an error inside without returning `Self`.